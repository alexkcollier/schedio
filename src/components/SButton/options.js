export {
  colors,
  allShadeOptions,
  colorNames,
} from '@/utils/component-options'

export const sizes = [
  'small',
  'regular',
  'large',
]

export const types = [
  'outlined',
  'text',
]
