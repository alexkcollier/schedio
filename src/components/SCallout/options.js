export const calloutTypes = ['', 'success', 'warning', 'error']
