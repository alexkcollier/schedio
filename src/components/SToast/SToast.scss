// component styles
.toast {
  align-items: center;
  background-color: color('white');
  border-radius: border-radius('small');
  bottom: spacing();
  box-shadow: shadow('border'), shadow();
  display: flex;
  flex: 0 0 auto;
  margin: spacing();
  margin-bottom: 0;
  max-width: 44em;
  pointer-events: auto;
  z-index: 500;

  &:first-child {
    margin-bottom: spacing();
  }

  &__icon {
    align-items: center;
    display: flex;
    padding: spacing();
  }

  &__text {
    padding: spacing();

    &:not(:first-child) {
      padding-left: 0;
    }

    &:not(:last-child) {
      padding-right: spacing('half');
    }
  }

  &__title {
    font-weight: $font-weight-bold;
    text-transform: uppercase;
  }

  &__action {
    align-self: stretch;
    border-bottom-right-radius: border-radius();
    border-top-right-radius: border-radius();
    background-color: transparent;
    color: color('spartan_blue');
    cursor: pointer;
    padding: spacing();

    &:hover {
      background-color: color('grey', 'lighter');
    }

    &:active {
      background-color: color('grey', 'light');
    }
  }

  &--warning {
    .toast__action {
      color: color('gold', 'dark');
    }
  }

  &--error {
    .toast__action {
      color: color('red');
    }
  }

  &--success {
    .toast__action {
      color: color('green', 'dark');
    }
  }

  &--right {
    align-self: flex-end;
  }

  &--center {
    align-self: center;
  }
}
