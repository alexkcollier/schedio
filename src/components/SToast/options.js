export const types = [
  '',
  'success',
  'warning',
  'error',
]
export const positions = [
  '',
  'center',
  'right',
]
