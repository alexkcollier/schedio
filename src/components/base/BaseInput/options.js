export const allowed = ['email', 'text', 'password', 'url', 'search', ''].sort()
