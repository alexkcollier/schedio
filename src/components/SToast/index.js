import Vue from 'vue'
import SToast from './SToast.vue'

const SToastProgrammatic = {
  open (params) {
    const vm = typeof window !== 'undefined' && window.Vue ? window.Vue : Vue
    const SToastComponent = vm.extend(SToast)
    const propsData = Object.assign({}, params)

    return new SToastComponent({
      el: document.createElement('div'),
      propsData,
    })
  },
}

export { SToast }

export default {
  install: (Vue) => {
    Vue.prototype.$toast = SToastProgrammatic
  },
}
