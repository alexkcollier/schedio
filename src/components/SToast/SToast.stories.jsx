import { SToast } from '@/components/SToast'
import { SHeading } from '@/components/SHeading'
import { SButton } from '@/components/SButton'
import PropList from '@/docs/PropList'
import StoryContainer from '@/docs/StoryContainer'
import { boolean, number, select, text, withKnobs } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/vue'
import { toastTypes, positions } from './options'

storiesOf('Components/SToast', module)
  .addDecorator(withKnobs)
  .add('SToast', () => {
    return {
      props: {
        props: {
          default: {
            toastType: select('toast-type', toastTypes, toastTypes[0], 'Required Props'),
            containerParent: text('container-parent', '#example', 'Optional Props'),
            toastBody: text('toast-body', 'Some text', 'Optional Props'),
            position: select('position', positions, positions[0], 'Optional Props'),
            actionText: text('action-text', 'Run an action', 'Optional Props'),
            toastTitle: text('toast-title', '', 'Optional Props'),
            hasIcon: boolean('has-icon', true, 'Optional Props'),
            isIndefinite: boolean('is-indefinite', false, 'Optional Props'),
            duration: number('duration', 3000, {}, 'Optional Props'),

            action() {
              setTimeout(() => {
                this.$toast.open({
                  containerParent: '#example',
                  toastBody: 'The action ran',
                  toastType: 'warning'
                })
              }, 750)
            }
          }
        }
      },

      render(h) {
        const { props } = this.$props
        const exampleContainerStyle = {
          minHeight: '12em',
          position: 'relative',
          padding: '1em',
          backgroundColor: '#f2f5f7'
        }

        return (
          <StoryContainer>
            <SHeading level="1">SToast</SHeading>
            <p>
              Toasts are quick notifications, usually indicating the status of an action. They can
              also be used to trigger another action. To do so, just pass a function to the{' '}
              <code>action</code> prop.
            </p>

            <SHeading level="2">Example</SHeading>

            <div id="example" style={exampleContainerStyle}>
              <SButton outline-color="night" onclick={() => this.$toast.open(props)}>
                Open a toast
              </SButton>
            </div>

            {SToast.props && <PropList component={SToast} />}
          </StoryContainer>
        )
      }
    }
  })
