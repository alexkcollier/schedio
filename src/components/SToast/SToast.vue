<template>
  <Transition name="fade">
    <div
      v-show="isVisible"
      :class="classList"
      class="toast"
    >
      <!-- Icon -->
      <div
        v-if="type && !hideIcon"
        class="toast__icon"
      >
        <SIcon
          :icon="toastIcon"
          :color="toastIconColor"
          aria-hidden="true"
        />
      </div>

      <!-- Content -->
      <div
        v-if="hasContent.any"
        class="toast__text"
      >
        <!-- Title -->
        <div
          v-if="hasContent.title"
          class="toast__title"
        >
          {{ title }}
        </div>

        <!-- Body -->
        <div
          v-if="hasContent.body"
          class="toast__body"
        >
          {{ body }}
        </div>
      </div>

      <button
        v-if="action && (typeof action === 'function')"
        class="toast__action"
        @click="clickHandler"
      >
        {{ actionText }}
      </button>
    </div>
  </Transition>
</template>

<script>
import NotificationMixin from '@/mixins/Notification.mixin';
import { SIcon } from '@/components/SIcon';
import { types, positions } from './options';

export default {
  name: 'SToast',

  components: {
    SIcon,
  },

  mixins: [NotificationMixin],

  props: {
    type: {
      type: String,
      default: '',
      validator: (v) => {
        return types.includes(v) || console.error(`\`type\` should be one of ${types.join(', ')}`);
      },
    },

    hideIcon: {
      type: Boolean,
      default: false,
    },

    title: {
      type: String,
      default: '',
    },

    body: {
      type: String,
      default: '',
    },

    action: {
      type: Function,
      default: null,
    },

    actionText: {
      type: String,
      default: 'OK',
    },

    position: {
      type: String,
      default: '',
      validator: (v) => {
        return (
          positions.includes(v) ||
          console.error(`\`position\` should be one of ${positions.join(', ')}`)
        );
      },
    },
  },

  computed: {
    classList () {
      return {
        [`toast--${this.type}`]: this.type,
        [`toast--${this.position}`]: this.position,
      };
    },

    toastIcon () {
      const stateMap = {
        error: 'alert-circle',
        warning: 'alert-triangle',
        success: 'check-circle',
      };

      return stateMap[this.type];
    },

    toastIconColor () {
      const stateMap = {
        error: 'red',
        warning: 'gold',
        success: 'green',
      };

      return stateMap[this.type];
    },

    hasContent () {
      const title = this.text || this.$slots.title;
      const body = this.body || this.$slots.default;
      const any = title || body;

      return { any, title, body };
    },
  },

  mounted () {
    if (!this.hasContent.any) console.error('`SToast` requires title or body text');
  },

  methods: {
    clickHandler () {
      this.action();
      this.closeNotification();
    },
  },
};
</script>

<style lang="scss">
.toast {
  align-items: center;
  background-color: color('white');
  border-radius: border-radius('small');
  bottom: spacing();
  box-shadow: shadow('border'), shadow();
  display: flex;
  flex: 0 0 auto;
  margin: spacing();
  max-width: 44rem;
  pointer-events: auto;
  z-index: 500;

  & + & {
    margin-bottom: spacing('none');
  }

  &__icon {
    align-items: center;
    display: flex;
    padding: spacing();
  }

  &__text {
    padding: spacing();

    &:not(:first-child) {
      padding-left: 0;
    }

    &:not(:last-child) {
      padding-right: spacing('half');
    }
  }

  &__title {
    font-weight: font-weight('bold');
    text-transform: uppercase;
  }

  &__action {
    align-self: stretch;
    background-color: transparent;
    border-bottom-right-radius: border-radius('small');
    border-top-right-radius: border-radius('small');
    color: color('blue');
    padding: spacing();

    &:hover {
      background-color: color('grey', 'lighter');
    }

    &:active {
      background-color: color('grey', 'light');
    }
  }

  &--warning {
    .toast__action {
      color: color('gold', 'dark');
    }
  }

  &--error {
    .toast__action {
      color: color('red');
    }
  }

  &--success {
    .toast__action {
      color: color('green', 'dark');
    }
  }

  &--right {
    align-self: flex-end;
  }

  &--center {
    align-self: center;
  }
}
</style>
