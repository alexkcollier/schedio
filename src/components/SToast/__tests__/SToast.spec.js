import SToast from '@/components/SToast/SToast.vue';
import { mount, shallowMount } from '@vue/test-utils';
import { positions, types } from '../options';
import { axe, toHaveNoViolations } from 'jest-axe';

expect.extend(toHaveNoViolations);

describe('SToast.vue', () => {
  const errorSpy = jest.spyOn(global.console, 'error').mockImplementation(() => { });
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(SToast, {
      propsData: {
        type: types[0],
        body: 'Text',
      },
    });
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  types.forEach((type) => {
    it(`renders toast type ${type} correctly`, async () => {
      jest.useFakeTimers();

      const tempWrapper = mount(SToast, {
        propsData: {
          type,
          body: 'Text',
        },
      });

      // workaround for async callback not running in expected timeframe
      // https://github.com/testing-library/react-testing-library/issues/244
      Promise.resolve().then(() => jest.runAllTimers());

      expect(tempWrapper.html()).toMatchSnapshot();
      expect(await axe(tempWrapper.html())).toHaveNoViolations();
    });
  });

  it('warns if the wrong toast type is provided', () => {
    shallowMount(SToast, {
      propsData: {
        type: 'not a type',
        body: 'Text',
      },
    });

    expect(errorSpy).toHaveBeenCalled();
  });

  it('requires a title or body', () => {
    const mountToast = (props) => {
      return shallowMount(SToast, {
        propsData: {
          type: types[0],
          ...(props && props),
        },
      });
    };

    mountToast();

    expect(errorSpy).toHaveBeenCalled();

    const body = mountToast({ body: 'text' });
    expect(body.html()).toMatchSnapshot();

    const title = mountToast({ title: 'title' });
    expect(title.html()).toMatchSnapshot();

    expect(errorSpy).toHaveBeenCalledTimes(2);
  });

  positions.forEach((position) => {
    it(`can have position ${position}`, async () => {
      wrapper.setProps({ position });
      await wrapper.vm.$nextTick();
      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  it('validates position', () => {
    shallowMount(SToast, {
      propsData: {
        type: types[0],
        body: 'Text',
        position: 'not a position',
      },
    });

    expect(errorSpy).toHaveBeenCalled();
  });

  it('can have a clickable action', async () => {
    const action = jest.fn();

    wrapper.setProps({ action });
    await wrapper.vm.$nextTick();
    // button should be visible on setting an action
    expect(wrapper.html()).toMatchSnapshot();

    wrapper.find('button').trigger('click');
    await wrapper.vm.$nextTick();
    expect(action).toHaveBeenCalled();
  });

  it('closes on action click', async () => {
    wrapper.setProps({ action: jest.fn() });
    await wrapper.vm.$nextTick();
    wrapper.find('button').trigger('click');
    await wrapper.vm.$nextTick();

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('can have action text configured', async () => {
    const actionText = 'action text';

    wrapper.setProps({
      action: jest.fn(),
      actionText,
    });

    await wrapper.vm.$nextTick();
    expect(wrapper.text()).toContain(actionText);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('can have the icon hidden', async () => {
    wrapper.setProps({ hideIcon: true });
    await wrapper.vm.$nextTick();
    expect(wrapper.html()).toMatchSnapshot();
  });
});
