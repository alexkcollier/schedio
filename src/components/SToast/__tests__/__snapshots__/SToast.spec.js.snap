// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SToast.vue can have a clickable action 1`] = `
<div class="toast fade-enter fade-enter-active" style="">
  <!---->
  <div class="toast__text">
    <!---->
    <div class="toast__body">
      Text
    </div>
  </div> <button class="toast__action">
    OK
  </button>
</div>
`;

exports[`SToast.vue can have action text configured 1`] = `
<div class="toast fade-enter fade-enter-active" style="">
  <!---->
  <div class="toast__text">
    <!---->
    <div class="toast__body">
      Text
    </div>
  </div> <button class="toast__action">
    action text
  </button>
</div>
`;

exports[`SToast.vue can have position  1`] = `
<div class="toast fade-enter fade-enter-active" style="">
  <!---->
  <div class="toast__text">
    <!---->
    <div class="toast__body">
      Text
    </div>
  </div>
  <!---->
</div>
`;

exports[`SToast.vue can have position center 1`] = `
<div class="toast toast--center fade-enter fade-enter-active" style="">
  <!---->
  <div class="toast__text">
    <!---->
    <div class="toast__body">
      Text
    </div>
  </div>
  <!---->
</div>
`;

exports[`SToast.vue can have position right 1`] = `
<div class="toast toast--right fade-enter fade-enter-active" style="">
  <!---->
  <div class="toast__text">
    <!---->
    <div class="toast__body">
      Text
    </div>
  </div>
  <!---->
</div>
`;

exports[`SToast.vue can have the icon hidden 1`] = `
<div class="toast fade-enter fade-enter-active" style="">
  <!---->
  <div class="toast__text">
    <!---->
    <div class="toast__body">
      Text
    </div>
  </div>
  <!---->
</div>
`;

exports[`SToast.vue closes on action click 1`] = `
<div class="toast fade-leave fade-leave-active" style="">
  <!---->
  <div class="toast__text">
    <!---->
    <div class="toast__body">
      Text
    </div>
  </div> <button class="toast__action">
    OK
  </button>
</div>
`;

exports[`SToast.vue renders toast type  correctly 1`] = `
<div class="toast fade-enter fade-enter-active" style="">
  <!---->
  <div class="toast__text">
    <!---->
    <div class="toast__body">
      Text
    </div>
  </div>
  <!---->
</div>
`;

exports[`SToast.vue renders toast type error correctly 1`] = `
<div class="toast toast--error fade-enter fade-enter-active" style="">
  <div class="toast__icon"><svg viewBox="0 0 24 24" class="icon icon--color-red">
      <circle cx="12" cy="12" r="10"></circle>
      <line x1="12" y1="8" x2="12" y2="12"></line>
      <line x1="12" y1="16" x2="12" y2="16"></line>
    </svg></div>
  <div class="toast__text">
    <!---->
    <div class="toast__body">
      Text
    </div>
  </div>
  <!---->
</div>
`;

exports[`SToast.vue renders toast type success correctly 1`] = `
<div class="toast toast--success fade-enter fade-enter-active" style="">
  <div class="toast__icon"><svg viewBox="0 0 24 24" class="icon icon--color-green">
      <path d="M22 11.08V12a10 10 0 1 1-5.93-9.14"></path>
      <polyline points="22 4 12 14.01 9 11.01"></polyline>
    </svg></div>
  <div class="toast__text">
    <!---->
    <div class="toast__body">
      Text
    </div>
  </div>
  <!---->
</div>
`;

exports[`SToast.vue renders toast type warning correctly 1`] = `
<div class="toast toast--warning fade-enter fade-enter-active" style="">
  <div class="toast__icon"><svg viewBox="0 0 24 24" class="icon icon--color-gold">
      <path d="M10.29 3.86L1.82 18a2 2 0 0 0 1.71 3h16.94a2 2 0 0 0 1.71-3L13.71 3.86a2 2 0 0 0-3.42 0z"></path>
      <line x1="12" y1="9" x2="12" y2="13"></line>
      <line x1="12" y1="17" x2="12" y2="17"></line>
    </svg></div>
  <div class="toast__text">
    <!---->
    <div class="toast__body">
      Text
    </div>
  </div>
  <!---->
</div>
`;

exports[`SToast.vue requires a title or body 1`] = `
<div class="toast fade-enter fade-enter-active" style="">
  <!---->
  <div class="toast__text">
    <!---->
    <div class="toast__body">
      text
    </div>
  </div>
  <!---->
</div>
`;

exports[`SToast.vue requires a title or body 2`] = `
<div class="toast fade-enter fade-enter-active" style="">
  <!---->
  <!---->
  <!---->
</div>
`;
